packages/entity-cache-adapter-redis/src/__integration-tests__/ShardedRedisCacheAdapter-integration-test.ts

```diff
@@ -0,0 +1,218 @@
+  return value;
+}
+
+type SetInstruction = { key: string; value: string; secondsToken: 'EX'; seconds: number };
+
+class ShardedRedis implements IRedis {
+  constructor(private readonly redium: readonly Redis[]) {}
+
+  async mget(...args: string[]): Promise<(string | null)[]> {
+    const originalKeys = args;
+    const redisClientsForKeys = this.getRedisClientsForKeys(args);
+    const allKeysToValues = new Map<string, string | null>();
+    await mapMapAsync(redisClientsForKeys, async ({ redisClient, keys }) => {
+      const values = await redisClient.mget(keys);
+      for (let i = 0; i < keys.length; i++) {
+        allKeysToValues.set(nullthrows(keys[i]), assertNotUndefined(values[i]));
+      }
+    });
+
+    return originalKeys.map((k) => assertNotUndefined(allKeysToValues.get(k)));
+  }
+  multi(): IRedisTransaction {
+    const accumulatedSets: Map<string, SetInstruction> = new Map();
+    const parentExec = async (): Promise<void> => {
+      const redisClientsForKeys = this.getRedisClientsForKeys([...accumulatedSets.keys()]);
+      const redisClientsForInstructions = mapMap(redisClientsForKeys, ({ keys, redisClient }) => {
+        return {
+          keys,
+          sets: keys.map((k) => assertNotUndefined(accumulatedSets.get(k))),
+          redisClient,
+        };
+      });
+
+      await mapMapAsync(redisClientsForInstructions, async ({ sets, redisClient }) => {
+        let redisTransaction = redisClient.multi();
+        sets.forEach(({ key, value, secondsToken, seconds }) => {
+          redisTransaction = redisTransaction.set(key, value, secondsToken, seconds);
+        });
+        await redisTransaction.exec();
+      });
+    };
+    const ret: IRedisTransaction = {
+      set(key, value, secondsToken, seconds): any {
+        accumulatedSets.set(key, { key, value, secondsToken, seconds });
+        return ret;
+      },
+      async exec(): Promise<any> {
+        return await parentExec();
+      },
+    };
+    return ret;
+  }
+
+  async del(...args: string[]): Promise<void> {
+    const redisClientsForKeys = this.getRedisClientsForKeys(args);
+    await mapMapAsync(redisClientsForKeys, async ({ keys, redisClient }) => {
+      await redisClient.del(...keys);
+    });
+  }
+
+  private getRedisClientsForKeys(
+    keys: readonly string[]
+  ): Map<number, { keys: string[]; redisClient: Redis }> {
+    const shardGroupsForKeys = new Map(keys.map((k) => [k, getShardForKey(k)]));
+    const redisClientsForKeys = new Map<number, { keys: string[]; redisClient: Redis }>();
+    for (const [key, shardGroup] of shardGroupsForKeys) {
+      const entry = computeIfAbsent(redisClientsForKeys, shardGroup, (currShardGroup) => ({
+        keys: [],
+        redisClient: this.getRedisInstanceForShardGroup(currShardGroup),
+      }));
+      entry.keys.push(key);
+    }
+    return redisClientsForKeys;
+  }
+
+  private getRedisInstanceForShardGroup(shardGroup: number): Redis {
```

Comment:
The other places in the code refer to "Redis clients":
```suggestion
  private getRedisClientForShardGroup(shardGroup: number): Redis {
```